cpp/src/buffer/buffer.cpp

```diff
@@ -139,8 +155,135 @@ std::unique_ptr<Buffer> Buffer::copy(MemoryType target, rmm::cuda_stream_view st
+    );
+    RAPIDSMPF_EXPECTS(
+        offset + std::ptrdiff_t(length) <= std::ptrdiff_t(size),
+        "offset + length can't be greater than size"
+    );
+
+    // Implement the copy between each possible memory types (both directions).
+    return std::visit(
+        overloaded{
+            [&](const HostStorageT& storage) {
+                switch (target_reserv.mem_type()) {
+                case MemoryType::HOST:  // host -> host
+                    return cuda_copy_slice_impl<cudaMemcpyHostToHost>(
+                        storage, offset, length, target_reserv, stream
+                    );
+                case MemoryType::DEVICE:  // host -> device
+                    return cuda_copy_slice_impl<cudaMemcpyHostToDevice>(
+                        storage, offset, length, target_reserv, stream
+                    );
+                }
+                RAPIDSMPF_FAIL("Invalid memory type");  // unreachable
+            },
+            [&](DeviceStorageT const& storage) {
+                switch (target_reserv.mem_type()) {
+                case MemoryType::HOST:  // device -> host
+                    return cuda_copy_slice_impl<cudaMemcpyDeviceToHost>(
+                        storage, offset, length, target_reserv, stream
+                    );
+                case MemoryType::DEVICE:  // device -> device
+                    return cuda_copy_slice_impl<cudaMemcpyDeviceToDevice>(
+                        storage, offset, length, target_reserv, stream
+                    );
+                }
+                RAPIDSMPF_FAIL("Invalid memory type");  // unreachable
+            }
+        },
+        storage_
+    );
+}
+
+std::ptrdiff_t Buffer::copy_to(
+    Buffer& dest, std::ptrdiff_t dest_offset, rmm::cuda_stream_view stream
+) const {
+    RAPIDSMPF_EXPECTS(
+        dest.size - size_t(dest_offset) >= size,
+        "destination buffer is too small",
+        std::invalid_argument
+    );
+
+    if (size == 0) {  // empty buffer, nothing to do
+        return 0;
+    }
+
+    cudaMemcpyKind kind;
+
+    // if both buffers are on the host, use memcpy, otherwise, use cudaMemcpyAsync
+    if (mem_type() == MemoryType::HOST && dest.mem_type() == MemoryType::HOST) {
+        // if the source buffer is not ready, return 0
+        if (!is_ready()) {
+            return 0;
+        }
+        // src buffer is ready, copy the data
+        std::memcpy(static_cast<uint8_t*>(dest.data()) + dest_offset, data(), size);
+        return std::ptrdiff_t(size);
+    } else if (mem_type() == MemoryType::HOST) {
+        kind = cudaMemcpyHostToDevice;
+    } else if (dest.mem_type() == MemoryType::HOST) {
+        kind = cudaMemcpyDeviceToHost;
+    } else {
+        kind = cudaMemcpyDeviceToDevice;
+    }
+
+    RAPIDSMPF_CUDA_TRY(cudaMemcpyAsync(
+        static_cast<cuda::std::byte*>(dest.data()) + dest_offset,
+        data(),
+        size,
+        kind,
+        stream
+    ));
+    return std::ptrdiff_t(size);
```

Comment:
Needs to `cudaEventRecord(stream)` for the destination buffer.

Reply:
Still relevant.

Reply:
@wence- Yes, but this was intentional. Say we want to concat a list of buffers, to a destination buffer. So, we can differ event creation after calling all `copy_to`s.

Reply:
ah sorry, I meant cudastreamewaitevent

Reply:
ie the copy on this stream has to be ordered wrt the work captured in the event of the source buffer